erpnext/accounts/doctype/budget/budget.py

```diff
@@ -14,20 +14,34 @@ class DuplicateBudgetError(frappe.ValidationError): pass
 
 class Budget(Document):
 	def autoname(self):
-		self.name = make_autoname(self.cost_center + "/" + self.fiscal_year + "/.###")
-
+		if self.cost_center:
+			self.name = make_autoname(self.cost_center + "/" + self.fiscal_year + "/.###")
+		elif self.project:
+			self.name = make_autoname(self.project + "/" + self.fiscal_year + "/.###")
+	
 	def validate(self):
-		self.validate_duplicate()
+		if not self.cost_center and not self.project:
+			frappe.throw(_("Budget should be allocated against either Cost Center or Project."))
+		if self.cost_center:
+			self.validate_duplicate("cost_center")
+		if self.project:
+			self.validate_duplicate("project")
```

Comment:
Call `self.validate_duplicate()` only once and set budget_against inside the function (there is already a if...else)